src/SpikeCore/SpikeCore/Irc/IrcClientBase.cs

```diff
@@ -1,24 +1,55 @@
 ﻿using System;
 using System.Collections.Generic;
+using System.Threading;
 using SpikeCore.Domain;
 
 namespace SpikeCore.Irc
 {
     public abstract class IrcClientBase: IIrcClient
     {
         private IEnumerable<string> _channelsToJoin;
+        protected string _host;
+        protected int _port;
+        protected string _nickname;
         protected bool _authenticate;
         protected string _password;
+        protected bool _userInitiatedDisconnect;
+        
         public WebHostCancellationTokenHolder WebHostCancellationTokenHolder { protected get; set; }
-
         public virtual Action<string> MessageReceived { get; set; }
         public virtual Action<PrivMessage> PrivMessageReceived { get; set; }
+        public virtual bool IsConnected { get; }
 
         public virtual void Connect(string host, int port, string nickname, IEnumerable<string> channelsToJoin, bool authenticate, string password)
         {            
             _channelsToJoin = channelsToJoin;
             _authenticate = authenticate;
             _password = password;
+
+            _host = host;
+            _port = port;
+            _nickname = nickname;
+            
+            Connect();
+        }
+
+        protected void HandleDisconnect(object sender, EventArgs e)
+        {      
+            UnwireEvents();
+            
+            if (!_userInitiatedDisconnect)
+            {
+                while (!IsConnected)
+                {
+                    Connect();
+
+                    if (!IsConnected)
+                    {
+                        // TODO - [kog@epiphanic.org 01/17/2019]: replace with a max retries + exponential backoff
+                        Thread.Sleep(30000);
```

Comment:
Consider replacing Thread.Sleep with something like https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim?view=netcore-2.2 which will allow you to wait for a timeout, or for a signal. That way calling Quit can signal the Semaphore to release early.

```
semaphore.Wait(TimeSpan.FromSeconds(30));
```

Reply:
Done and done. This won't immediately be useful since we can't receive quit messages while we're disconnected, but it'll probably come in handy when we improve the ASP.NET front-end. Assuming we ever do that ;)